File: packages/sdk-go/verify.go
package queue

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// AdmitTokenClaims are the HS256 JWT claims returned on admit.
type AdmitTokenClaims map[string]any

// VerifyAdmitToken verifies an HS256 admit JWT with the same secret as security.jwtHmacSecret.
// Returns nil when the token is missing, malformed, expired, or signed with the wrong secret.
func VerifyAdmitToken(token, secret string, now time.Time) AdmitTokenClaims {
	if token == "" || secret == "" {
		return nil
	}
	parts := strings.Split(token, ".")
	if len(parts) != 3 || parts[0] == "" || parts[1] == "" || parts[2] == "" {
		return nil
	}
	headerJSON, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil
	}
	payloadJSON, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil
	}
	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return nil
	}

	var header struct {
		Alg string `json:"alg"`
	}
	if err := json.Unmarshal(headerJSON, &header); err != nil {
		return nil
	}
	if header.Alg != "" && header.Alg != "HS256" {
		return nil
	}

	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(parts[0] + "." + parts[1]))
	expected := mac.Sum(nil)
	if subtle.ConstantTimeCompare(expected, sig) != 1 {
		return nil
	}

	var claims AdmitTokenClaims
	if err := json.Unmarshal(payloadJSON, &claims); err != nil {
		return nil
	}
	if exp, ok := claims["exp"].(float64); ok {
		if now.IsZero() {
			now = time.Now()
		}
		if float64(now.Unix()) >= exp {
			return nil
		}
	}
	return claims
}

// ExtractAdmitToken reads vazue_token from a Cookie header or query values.
func ExtractAdmitToken(cookieHeader string, query url.Values) string {
	const name = "vazue_token"
	if cookieHeader != "" {
		for _, part := range strings.Split(cookieHeader, ";") {
			part = strings.TrimSpace(part)
			if k, v, ok := strings.Cut(part, "="); ok && k == name {
				if decoded, err := url.QueryUnescape(v); err == nil {
					return decoded
				}
				return v
			}
		}
	}
	if query != nil {
		if v := query.Get(name); v != "" {
			return v
		}
	}
	return ""
}

// ExtractAdmitTokenFromRequest is a convenience wrapper for http.Request.
func ExtractAdmitTokenFromRequest(r *http.Request) string {
	if r == nil {
		return ""
	}
	return ExtractAdmitToken(r.Header.Get("Cookie"), r.URL.Query())
}

File: packages/sdk-typescript/src/verify.ts
import { createHmac, timingSafeEqual } from 'node:crypto';

export type AdmitTokenClaims = {
  /** Visitor request id */
  sub?: string;
  tenant_id?: string;
  event_id?: string;
  request_id?: string;
  return_url?: string;
  exp?: number;
  [key: string]: unknown;
};

function b64urlToBuf(input: string): Buffer {
  const pad = '='.repeat((4 - (input.length % 4)) % 4);
  const b64 = (input + pad).replace(/-/g, '+').replace(/_/g, '/');
  return Buffer.from(b64, 'base64');
}

/**
 * Verify an HS256 admit JWT at the origin (Node 20+).
 * Use the same secret as `security.jwtHmacSecret` / data-plane signing.
 */
export function verifyAdmitToken(
  token: string,
  secret: string,
  nowSec = Math.floor(Date.now() / 1000),
): AdmitTokenClaims | null {
  if (!token || !secret) return null;
  const parts = token.split('.');
  if (parts.length !== 3 || parts.some((p) => !p)) return null;
  const [headerB64, payloadB64, sigB64] = parts;
  let header: { alg?: string };
  let payload: AdmitTokenClaims;
  try {
    header = JSON.parse(b64urlToBuf(headerB64).toString('utf8')) as { alg?: string };
    payload = JSON.parse(b64urlToBuf(payloadB64).toString('utf8')) as AdmitTokenClaims;
  } catch {
    return null;
  }
  if (header.alg && header.alg !== 'HS256') return null;

  const expected = createHmac('sha256', secret).update(`${headerB64}.${payloadB64}`).digest();
  let actual: Buffer;
  try {
    actual = b64urlToBuf(sigB64);
  } catch {
    return null;
  }
  if (expected.length !== actual.length || !timingSafeEqual(expected, actual)) {
    return null;
  }
  if (typeof payload.exp === 'number' && nowSec >= payload.exp) {
    return null;
  }
  return payload;
}

/** Read `vazue_token` from a Cookie header or query string. */
export function extractAdmitToken(opts: {
  cookieHeader?: string | null;
  cookieName?: string;
  query?: URLSearchParams | Record<string, string> | null;
}): string | undefined {
  const name = opts.cookieName ?? 'vazue_token';
  if (opts.cookieHeader) {
    for (const part of opts.cookieHeader.split(';')) {
      const [k, ...rest] = part.trim().split('=');
      if (k === name) return decodeURIComponent(rest.join('=') || '');
    }
  }
  if (opts.query) {
    if (opts.query instanceof URLSearchParams) return opts.query.get(name) ?? undefined;
    return opts.query[name];
  }
  return undefined;
}
